Bugzilla/ModPerl/BlockIP.pm

```diff
@@ -0,0 +1,59 @@
+package Bugzilla::ModPerl::BlockIP;
+use 5.10.1;
+use strict;
+use warnings;
+
+use Apache2::RequestRec ();
+use Apache2::Connection ();
+
+use Apache2::Const -compile => qw(OK);
+use Cache::Memcached::Fast;
+
+use constant BLOCK_TIMEOUT => 60*60;
+
+my $MEMCACHED = Bugzilla::Memcached->_new()->{memcached};
+my $STATIC_URI = qr{
+    ^/
+     (?: extensions/[^/]+/web
+       | robots\.txt
+       | __heartbeat__
+       | __lbheartbeat__
+       | __version__
+       | images
+       | skins
+       | js
+       | errors
+     )
+}xs;
+
+sub block_ip {
+    my ($class, $ip) = @_;
+    $MEMCACHED->set("block_ip:$ip" => 1, BLOCK_TIMEOUT) if $MEMCACHED;
+}
+
+sub unblock_ip {
+    my ($class, $ip) = @_;
+    $MEMCACHED->delete("block_ip:$ip") if $MEMCACHED;
+}
+
+sub handler {
+    my $r = shift;
+    return Apache2::Const::OK if $r->uri =~ $STATIC_URI;
+
+    my $ip = $r->headers_in->{'X-Forwarded-For'} // $r->connection->remote_ip;
```

Comment:
`X-Forwarded-For` can be forged by the client; they could set it to a different value with each request and bypass this block.

assuming there's always only one proxy server in front of bmo, you should be able to split on `[\s,]+` and use the penultimate element as the ip.

Reply:
Is that still the case? That would be a gross misconfiguration of Zeus.

Reply:
if there's proxy servers between the client and zeus that should appear in `x-forwarded-for`; zeus should just be appending its ip on the list.